spring-cloud-gcp-autoconfigure/src/test/java/org/springframework/cloud/gcp/autoconfigure/pubsub/it/PubSubChannelAdaptersIntegrationTests.java

```diff
@@ -0,0 +1,289 @@
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.integration.annotation.ServiceActivator;
+import org.springframework.integration.channel.QueueChannel;
+import org.springframework.integration.config.EnableIntegration;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageChannel;
+import org.springframework.messaging.MessageHeaders;
+import org.springframework.messaging.PollableChannel;
+import org.springframework.messaging.converter.StringMessageConverter;
+import org.springframework.messaging.support.MessageBuilder;
+import org.springframework.test.context.junit4.SpringRunner;
+import org.springframework.util.concurrent.ListenableFutureCallback;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assumptions.assumeThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+/**
+ * @author João André Martins
+ */
+@SpringBootTest(
+		properties = {"spring.cloud.gcp.pubsub.emulatorHost=${PUBSUB_EMULATOR_HOST}"},
+		classes = {PubSubChannelAdaptersIntegrationTests.IntegrationConfiguration.class,
+				GcpPubSubEmulatorConfiguration.class,
+				GcpPubSubAutoConfiguration.class
+		}
+)
+@RunWith(SpringRunner.class)
+public class PubSubChannelAdaptersIntegrationTests {
+
+	private static final String EMULATOR_HOST_ENVVAR_NAME = "PUBSUB_EMULATOR_HOST";
+
+	@Autowired
+	@Qualifier("outputChannel")
+	private PollableChannel channel;
+
+	@Autowired
+	private PubSubAdmin pubSubAdmin;
+
+	@Autowired
+	private GcpPubSubProperties pubSubProperties;
+
+	@Autowired
+	@Qualifier("inputChannel")
+	private MessageChannel inputChannel;
+
+	@Autowired
+	private PubSubInboundChannelAdapter inboundChannelAdapter;
+
+	@Autowired
+	private PubSubMessageHandler outboundChannelAdapter;
+
+	// Increment for each added test.
+	// Used in a JUnit5 @AfterAll way, to clean up the Pub/Sub environment after all tests ran.
+	// Can't use @AfterClass, since the PubSubAdmin bean isn't ready at static time.
+	private static final int TOTAL_TESTS = 4;
+
+	private static int testsRan;
+
+	@BeforeClass
+	public static void checkEmulatorIsRunning() {
+		assumeThat(System.getenv(EMULATOR_HOST_ENVVAR_NAME)).isNotNull();
+	}
+
+	@Before
+	public void setUp() {
+		// Sets the defaults of the fields we'll change in later tests, so we don't need to use
+		// @DirtiesContext, which is performance unfriendly.
+		StringMessageConverter stringMessageConverter = new StringMessageConverter();
+		stringMessageConverter.setSerializedPayloadClass(String.class);
+		this.inboundChannelAdapter.setMessageConverter(stringMessageConverter);
+		this.inboundChannelAdapter.setAckMode(AckMode.AUTO);
+		this.outboundChannelAdapter.setPublishCallback(null);
+	}
+
+	@After
+	public void tearDown() {
+		if (testsRan == TOTAL_TESTS) {
```

Comment:
Can't we really have a `PubSubAdmin` as a statically configured and really rely on the `@AfterClass` instead?

I don't tell that we should somehow worry about subsequent application context we use (although we can use `static` properties for beans as well), but that way we won't forget what's going on with this hack when we add more tests.

Another way to go is just only have one `@Test` in the class.